Upgrade to React 18

renchap opened the pull request:
This upgrades React to the latest version. The upgrade is easy, thanks to React's policy of supporting old APIs for a long time.

Official docs:
- https://legacy.reactjs.org/blog/2020/10/20/react-v17.html
- https://react.dev/blog/2022/03/08/react-18-upgrade-guide

First commit renames the legacy class component methods (using the official codemod).

The second upgrade is the real update, with the changes to render components using the new API.

Please note that some libraries are not showing React 18 compat in their `package.json`. Most of those are deprecated and React's API did not change, but this displays warnings in console (they were already displayed with the current React 16 version):
```
emoji-mart@3.0.1-j
react-intl@2.9.0
react-motion@0.5.2
react-notification@6.8.5
react-router-scroll-4@1.0.0-beta.2
react-swipeable-views@0.14.0
```

Most of those are deprecated or no longer maintained, so it will be a non-trivial feature to update them.

A possible issue will be the [change in event delegation](https://legacy.reactjs.org/blog/2020/10/20/react-v17.html#changes-to-event-delegation), React events are no longer attached to `document`, but to the React root DOM element. I am not sure if this has an impact in the codebase, and there is no way to really test it.

I choose to *not* enable the [new JSX transform](https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html) as this will create a lot of changes.
It brings some perf improvements and remove the need for `React` to be imported in every JSX file), so I intend to do it in another PR once this one is merged.

renchap committed:
Upgrade to React 18

Using the new `createRoot` API

renchap commented:
@Gargron @ClearlyClaire @ykzts  Do you see any reason not to do this?

I marked the PR as draft for now as I would like your approvals, but it ran correctly locally. We dont have many tests covering the frontend, so I am not sure how to ensure it has no issue.

ykzts commented:
Duplicates work on #18306, but #18306 is WIP because conflict resolution is stalled. I am closing #18306 as it is now very out of date.

nschonni commented:
Probably makes sense to turn on this rule for this PR https://github.com/mastodon/mastodon/pull/24471

ClearlyClaire commented:
> @Gargron @ClearlyClaire @ykzts Do you see any reason not to do this?

I have nothing against this change based on your description, though this needs some testing.

> I marked the PR as draft for now as I would like your approvals, but it ran correctly locally. We dont have many tests covering the frontend, so I am not sure how to ensure it has no issue.

I'll test this when I have some time. I think it should be mostly fine, but the event delegation changes might possibly cause issues. One thing I'd check in particular is the moderation/admin UI which uses React very sparingly, but I think it should be fine too.

ClearlyClaire commented:
This breaks at least the dropdown menus in the web app (I have not investigated why yet).

I could not test the media attachments in the moderation interface as this is broken for another reason (that I will investigate next).

ClearlyClaire commented:
Dropdowns are still broken, I haven't investigated what is breaking them yet.

renchap commented:
> I am not quite sure how that is happening, but it seems the issue is that the click event causes the dropdown menu to be mounted, add its `handleDocumentClick` listener, _and_ process the click event for that same click, causing it to immediately close.

I will have a look, is it a specific place for me to look at?

renchap commented:
This should be fixable by https://legacy.reactjs.org/blog/2020/08/10/react-v17-rc.html#fixing-potential-issues

I tested it locally and it works, but I need to check every place that relies on this.

renchap committed:
Fix event delegation with React 17+

Documentation: https://legacy.reactjs.org/blog/2020/08/10/react-v17-rc.html#fixing-potential-issues

The recommended fix is to use `{ capture: true }` when defining the events.

I went through the various `document.addEventListener` calls and updated the ones that I found were not working.

If the `options` parameter is a boolean, this is used as the `capture` option.

I also unified the various calls to handle browsers with passive events the same way as browsers not supporting it.

renchap commented:
@ClearlyClaire Should be fixed, see my last commit

renchap committed:
Fix Dropdowns not closing when button is clicked again

renchap commented:
This is now fixed. As `capture` is set on the event handlers, `handleDocumentClick` was called, which closed the menu, but then `handleClick` was called, which opened it.

The solution is to stop propagating the event.

ClearlyClaire commented on app/javascript/mastodon/components/dropdown_menu.jsx:
```diff
@@ -35,6 +35,7 @@ class DropdownMenu extends React.PureComponent {
   handleDocumentClick = e => {
     if (this.node && !this.node.contains(e.target)) {
       this.props.onClose();
+      e.stopPropagation();
```
Isn't that inconsistent with the `passive: true` option used on `touchend`?

renchap replied:
`passive` is about `preventDefault()`,  not `stopPropagation()`, according to MDN

ClearlyClaire replied:
You're right, that makes sense. Looks good then.

ClearlyClaire commented on app/javascript/mastodon/features/compose/components/language_dropdown.jsx:
```diff
@@ -39,6 +39,7 @@ class LanguageDropdownMenu extends React.PureComponent {
   handleDocumentClick = e => {
     if (this.node && !this.node.contains(e.target)) {
       this.props.onClose();
+      e.stopPropagation();
```
Isn't that inconsistent with the `passive: true` option used on `touchend`?

ClearlyClaire commented on app/javascript/mastodon/features/compose/components/privacy_dropdown.jsx:
```diff
@@ -34,6 +34,7 @@ class PrivacyDropdownMenu extends React.PureComponent {
   handleDocumentClick = e => {
     if (this.node && !this.node.contains(e.target)) {
       this.props.onClose();
+      e.stopPropagation();
```
Isn't that inconsistent with the `passive: true` option used on `touchend`?

ClearlyClaire commented on app/javascript/mastodon/components/column.jsx:
```diff
@@ -3,6 +3,8 @@ import PropTypes from 'prop-types';
 import { supportsPassiveEvents } from 'detect-passive-events';
 import { scrollTop } from '../scroll';
 
+const listenerOptions = supportsPassiveEvents ? { passive: true } : true;
```
Shouldn't it use `capture: true` in both cases?
```suggestion
const listenerOptions = supportsPassiveEvents ? { passive: true, capture: true } : true;
```

That being said, I'm not sure this class needs changes at all.

renchap replied:
I chose to move the inline ones to `listenerOptions` here, to be similar to the other files (and its ran only one time).

The mistake here is that when there is no passive events support, the argument should be `false` (or `undefined`, `false` is the default value), which is consistent with the object arg.

ClearlyClaire commented on app/javascript/mastodon/components/scrollable_list.jsx:
```diff
@@ -15,6 +15,8 @@ import { connect } from 'react-redux';
 
 const MOUSE_IDLE_DELAY = 300;
 
+const listenerOptions = supportsPassiveEvents ? { passive: true } : true;
```
Exactly the same remarks as for `app/javascript/mastodon/components/column.jsx`

ClearlyClaire commented:
I have made some inline comments for stuff I found weird, but I have seen no actual runtime issue when testing (I haven't tested on a real mobile device though).

renchap committed:
Fix `capture` argument value when no support for capture events

renchap commented:
A big question is: do we want to keep `detect-passive-events` at all? Passive events are supported in most browsers now: https://caniuse.com/passive-event-listener
I think we would remove it everywhere to simplify the code, and use the object form of `options` directly.

I pushed a new commit with the issues you caught (which would have occurred only on old browsers, which probably break before that :D

nschonni commented:
Yarn is spitting out a few warnings about incompatible versions for some of the plugins
```
warning " > emoji-mart@3.0.1-j" has incorrect peer dependency "react@^0.14.0 || ^15.0.0-0 || ^16.0.0 || ^17.0.0".
warning " > react-intl@2.9.0" has incorrect peer dependency "react@^0.14.9 || ^15.0.0 || ^16.0.0".
warning " > react-motion@0.5.2" has incorrect peer dependency "react@^0.14.9 || ^15.3.0 || ^16.0.0".
warning " > react-notification@6.8.5" has incorrect peer dependency "react@^0.14.0 || ^15.0.0 || ^16.0.0".
warning " > react-router-scroll-4@1.0.0-beta.2" has incorrect peer dependency "react@^15.0.0 || ^16.0.0".
warning " > react-router-scroll-4@1.0.0-beta.2" has incorrect peer dependency "react-dom@^15.0.0 || ^16.0.0".
warning " > react-swipeable-views@0.14.0" has incorrect peer dependency "react@^15.3.0 || ^16.0.0 || ^17.0.0".
warning "react-swipeable-views > react-swipeable-views-utils > react-event-listener@0.6.6" has incorrect peer dependency "react@^16.3.0".
```

renchap commented:
Yes, this is unfortunate but all those libraries (except `react-intl`) are no longer maintained and never updated their `peerDependencies` to include modern React versions.

They most probably need to be replaced by more maintained alternatives, this is on my list…
